internal/provider/data_source_secrets_test.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package provider
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-testing/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	internaltesting "github.com/juju/terraform-provider-juju/internal/testing"
+)
+
+func TestAcc_DataSourceSecret(t *testing.T) {
+	if os.Getenv("JUJU_AGENT_VERSION") == "" || internaltesting.CompareVersions(os.Getenv("JUJU_AGENT_VERSION"), "3.3.0") < 0 {
+		t.Skip("JUJU_AGENT_VERSION is not set or is below 3.3.0")
+	}
+	modelName := acctest.RandomWithPrefix("tf-datasource-secret-test-model")
+	// ...-test-[0-9]+ is not a valid secret name, need to remove the dash before numbers
+	secretName := fmt.Sprintf("tf-datasource-secret-test%d", acctest.RandInt())
+	secretValue := map[string]string{
+		"key1": "value1",
+		"key2": "value2",
+	}
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:                 func() { testAccPreCheck(t) },
+		ProtoV6ProviderFactories: frameworkProviderFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceSecret(modelName, secretName, secretValue),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("data.juju_secret.secret_data_source", "model", modelName),
+					resource.TestCheckResourceAttr("data.juju_secret.secret_data_source", "name", secretName),
+					resource.TestCheckResourceAttrPair("data.juju_secret.secret_data_source", "secretID", "juju_secret.secret_resource", "secretID"),
+				),
+			},
+		},
+	})
+}
+
+func testAccDataSourceSecret(modelName, secretName string, secretValue map[string]string) string {
+	return internaltesting.GetStringFromTemplateWithData(
+		"testAccResourceSecret",
+		`
+resource "juju_model" "{{.ModelName}}" {
+  name = "{{.ModelName}}"
+}
+
+resource "juju_secret" "secret_resource" {
+  model = juju_model.{{.ModelName}}.name
+  name  = "{{.SecretName}}"
+  value =  {
+    {{- range $key, $value := .SecretValue }}
+    "{{$key}}" = "{{$value}}"
+    {{- end }}
+  }
+}
+
+data "juju_secret" "secret_data_source" {
+  name = juju_secret.secret_resource.name
+  model = juju_model.{{.ModelName}}.name
```

Comment:
This test is a bit contrived.  But we need a good charm for this and full implementation of all 3 pieces for secrets to do that. 

Please add a ToDo.

Reply:
Added, there is also one at the top of this PR description, am not planning on merging until its done